### ISSUE TITLE
Github bot account that auto-creates merge PRs between branches for various dotnet/* repo is hitting account limit.

### ISSUE THREAD

**Issue by vatsalyaagrawal** [dotnet/arcade#3151]
Problem description (by @brettfo) 

tl;dr – We need GitHub API rate limits removed for two bot accounts, dotnet-bot and dotnet-automerge-bot.

The problem:
The various dotnet/* repos on GitHub have a bot account that auto-creates merge PRs between branches (e.g., `master` to `release/dev16.3`, etc.), but we recently ran into failures because GitHub was rate-limiting the account.  To get around this a 5s delay was added between requests, but a consequence of this is that F# didn’t get auto-merges for over a week because we were near the bottom of the list and the function was being terminated after the default 5 minute timeout.  I was able to increase the timeout to the max of 10 minutes, and while this allowed the script to run to completion, we were only 9 seconds shy of the mark; we won’t be so lucky in the future.

The ask (ideal):
Can we get the API rate limit removed for the dotnet-bot and dotnet-automerge-bot accounts?  This will allow our regular merge bot to finish in ~4 minutes instead of 10 by removing the sleeps.

The ask (fall-back):
If the rate limit can’t be removed, how do we upgrade our roslyn-infra-functions account to whatever paid tier doesn’t have a 10 minute time limit?

What we really don’t want to do:
The final alternative is to split the merge tool into numerous subtools; e.g., one for Roslyn, FSharp, etc.  The downside to this is that we now have to look in multiple locations for merge bot errors.

**Comment by markwilkie** [comment]
cc/ @MattGal @chcosta

**Comment by chcosta** [comment]
I've reached out to GitHub folks on Teams to try to determine our options here.

**Comment by mmitche** [comment]
@vatsalyaagrawal @chcosta Do you know what rate limit is being hit? Is it the abuse limit or the hourly limit?

If it's the former, then it's unlikely to get removed, but the response returned by the API indicates when the request can be retried.

**Comment by markwilkie** [comment]
Given that this issue is old, I'm assuming it's been resolved.  Please reopen if not.

**Comment by brettfo** [comment]
I have not heard anything back about fixing the rate limit.  We're currently able to run our internal tool, but we're very close to hitting the rate limit on one end and the Azure Function timeout on the other.  As soon as we add more repos or branches to auto-merge then we'll start failing again.

**Comment by markwilkie** [comment]
Adding to FR.  cc/ @MattGal

**Comment by mmitche** [comment]
> Do you know what rate limit is being hit? Is it the abuse limit or the hourly limit?If it's the former, then it's unlikely to get removed, but the response returned by the API indicates when the request can be retried.

@brettfo, did you get a chance to look at what exceptions are being thrown?

**Comment by brettfo** [comment]
Our current logging is terrible around that; give me a bit to fix that and report back with what I find.

**Comment by MattGal** [comment]
@brettfo did we make any progress here?  Since noone's truly blocked and it's been open for months I'm thinking of closing it out, but we can keep it open if there's work to do.

**Comment by brettfo** [comment]
I haven't had an opportunity to dig in and since the tool is currently passing I'll close the issue.

If we start hitting it again I'll re-open this once I know the nature of the throttling.

**Comment by MattGal** [comment]
Thanks Brett!  I think the only escape valve we're likely to have to start creating more Apps/ bots if and when we need to.
